File: go/tests/hosting_mcp_test.go
// hosting_mcp_test.go
//
// Verifies CircleAI.Hosting.Mcp ports: the JSON-RPC 2.0 dispatcher (initialize,
// tools/list, tools/call, resources/list, resources/read), tool-level error
// envelopes, unknown-method / unknown-tool errors, and batch handling.

package circleai_test

import (
	"context"
	"encoding/json"
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
)

// echoMcpTool concatenates a fixed prefix with an argument.
type echoMcpTool struct{}

func (echoMcpTool) Name() string        { return "echo" }
func (echoMcpTool) Description() string { return "Echoes text" }
func (echoMcpTool) InputSchema() interface{} {
	return map[string]interface{}{"type": "object"}
}
func (echoMcpTool) Execute(_ context.Context, args map[string]interface{}) (interface{}, error) {
	return map[string]interface{}{"echoed": args["text"]}, nil
}

// erroringMcpTool always signals a tool-level error.
type erroringMcpTool struct{}

func (erroringMcpTool) Name() string             { return "boom" }
func (erroringMcpTool) Description() string      { return "Always fails" }
func (erroringMcpTool) InputSchema() interface{} { return map[string]interface{}{} }
func (erroringMcpTool) Execute(context.Context, map[string]interface{}) (interface{}, error) {
	return nil, circleai.NewMcpToolError("nope")
}

// memResourceProvider serves one vault:// resource.
type memResourceProvider struct{}

func (memResourceProvider) URIScheme() string { return "vault://" }
func (memResourceProvider) List(context.Context) ([]circleai.McpResource, error) {
	return []circleai.McpResource{{URI: "vault://note/1", Name: "Note 1", MimeType: "text/plain"}}, nil
}
func (memResourceProvider) Read(_ context.Context, uri string) (*circleai.McpResourceContent, error) {
	if uri == "vault://note/1" {
		return &circleai.McpResourceContent{URI: uri, MimeType: "text/plain", Text: "hello"}, nil
	}
	return nil, nil
}

func dispatch(t *testing.T, reg circleai.McpRegistry, req string) map[string]interface{} {
	t.Helper()
	resp, err := circleai.DispatchMcp(context.Background(), json.RawMessage(req), reg, circleai.DefaultMcpServerInfo())
	if err != nil {
		t.Fatalf("dispatch err: %v", err)
	}
	if resp == nil {
		return nil
	}
	// Round-trip through JSON to inspect as a generic map.
	b, _ := json.Marshal(resp)
	var m map[string]interface{}
	_ = json.Unmarshal(b, &m)
	return m
}

func TestMcp_Initialize(t *testing.T) {
	reg := circleai.McpRegistry{}
	m := dispatch(t, reg, `{"jsonrpc":"2.0","id":1,"method":"initialize"}`)
	result, ok := m["result"].(map[string]interface{})
	if !ok {
		t.Fatalf("no result: %v", m)
	}
	if result["protocolVersion"] != "2024-11-05" {
		t.Errorf("protocolVersion = %v", result["protocolVersion"])
	}
	// id is echoed as its JSON text form ("1").
	if m["id"] != "1" {
		t.Errorf("id = %v, want \"1\"", m["id"])
	}
}

func TestMcp_ToolsListAndCall(t *testing.T) {
	reg := circleai.McpRegistry{Tools: []circleai.IMcpTool{echoMcpTool{}}}

	list := dispatch(t, reg, `{"jsonrpc":"2.0","id":2,"method":"tools/list"}`)
	result := list["result"].(map[string]interface{})
	tools := result["tools"].([]interface{})
	if len(tools) != 1 || tools[0].(map[string]interface{})["name"] != "echo" {
		t.Fatalf("tools/list wrong: %v", tools)
	}

	call := dispatch(t, reg, `{"jsonrpc":"2.0","id":3,"method":"tools/call","params":{"name":"echo","arguments":{"text":"hi"}}}`)
	callResult := call["result"].(map[string]interface{})
	if callResult["isError"] != false {
		t.Errorf("isError = %v, want false", callResult["isError"])
	}
	content := callResult["content"].([]interface{})
	text := content[0].(map[string]interface{})["text"].(string)
	if text != `{"echoed":"hi"}` {
		t.Errorf("tool text = %q", text)
	}
}

func TestMcp_ToolError(t *testing.T) {
	reg := circleai.McpRegistry{Tools: []circleai.IMcpTool{erroringMcpTool{}}}
	m := dispatch(t, reg, `{"jsonrpc":"2.0","id":4,"method":"tools/call","params":{"name":"boom"}}`)
	result := m["result"].(map[string]interface{})
	if result["isError"] != true {
		t.Errorf("isError = %v, want true", result["isError"])
	}
	text := result["content"].([]interface{})[0].(map[string]interface{})["text"]
	if text != "nope" {
		t.Errorf("error text = %v", text)
	}
}

func TestMcp_UnknownTool(t *testing.T) {
	reg := circleai.McpRegistry{Tools: []circleai.IMcpTool{echoMcpTool{}}}
	m := dispatch(t, reg, `{"jsonrpc":"2.0","id":5,"method":"tools/call","params":{"name":"ghost"}}`)
	errObj := m["error"].(map[string]interface{})
	if errObj["code"].(float64) != -32602 {
		t.Errorf("code = %v, want -32602", errObj["code"])
	}
}

func TestMcp_UnknownMethod(t *testing.T) {
	m := dispatch(t, circleai.McpRegistry{}, `{"jsonrpc":"2.0","id":6,"method":"does/not/exist"}`)
	errObj := m["error"].(map[string]interface{})
	if errObj["code"].(float64) != -32601 {
		t.Errorf("code = %v, want -32601", errObj["code"])
	}
}

func TestMcp_Notification_ReturnsNil(t *testing.T) {
	resp, err := circleai.DispatchMcp(context.Background(),
		json.RawMessage(`{"jsonrpc":"2.0","method":"notifications/initialized"}`),
		circleai.McpRegistry{}, circleai.DefaultMcpServerInfo())
	if err != nil {
		t.Fatalf("dispatch: %v", err)
	}
	if resp != nil {
		t.Errorf("notification should return nil, got %v", resp)
	}
}

func TestMcp_ResourcesListAndRead(t *testing.T) {
	reg := circleai.McpRegistry{Resources: []circleai.IMcpResourceProvider{memResourceProvider{}}}

	list := dispatch(t, reg, `{"jsonrpc":"2.0","id":7,"method":"resources/list"}`)
	resources := list["result"].(map[string]interface{})["resources"].([]interface{})
	if len(resources) != 1 {
		t.Fatalf("resources/list = %v", resources)
	}
	first := resources[0].(map[string]interface{})
	if first["uri"] != "vault://note/1" || first["description"] != "Note 1" {
		t.Errorf("resource = %v (description should default to Name)", first)
	}

	read := dispatch(t, reg, `{"jsonrpc":"2.0","id":8,"method":"resources/read","params":{"uri":"vault://note/1"}}`)
	contents := read["result"].(map[string]interface{})["contents"].([]interface{})
	text := contents[0].(map[string]interface{})["text"]
	if text != "hello" {
		t.Errorf("read text = %v", text)
	}

	// Unknown scheme → error.
	noProv := dispatch(t, reg, `{"jsonrpc":"2.0","id":9,"method":"resources/read","params":{"uri":"models://x"}}`)
	if _, ok := noProv["error"]; !ok {
		t.Error("expected error for unknown scheme")
	}
}

func TestMcp_Batch(t *testing.T) {
	reg := circleai.McpRegistry{Tools: []circleai.IMcpTool{echoMcpTool{}}}
	body := `[{"jsonrpc":"2.0","id":1,"method":"initialize"},{"jsonrpc":"2.0","method":"notifications/initialized"},{"jsonrpc":"2.0","id":2,"method":"tools/list"}]`
	resp, err := circleai.DispatchMcpBatch(context.Background(), []byte(body), reg, circleai.DefaultMcpServerInfo())
	if err != nil {
		t.Fatalf("batch: %v", err)
	}
	arr, ok := resp.([]interface{})
	if !ok {
		t.Fatalf("batch result not an array: %T", resp)
	}
	// Notification produces no response → 2 responses, not 3.
	if len(arr) != 2 {
		t.Errorf("batch responses = %d, want 2 (notification omitted)", len(arr))
	}
}

File: rust/tests/hosting_mcp_test.rs
//! hosting_mcp_test.rs
//!
//! Verifies the MCP JSON-RPC 2.0 dispatcher: initialize, tools/list, tools/call
//! (success + tool error), resources/list, resources/read (found + not found +
//! no provider), method-not-found, notifications, and batch handling. Mirrors
//! the C# McpEndpoints.DispatchAsync.

use serde_json::json;

use circle_ai::hosting_mcp::{
    FnMcpTool, InMemoryResourceProvider, McpRegistry, McpServerInfo, McpToolError,
};

fn registry() -> McpRegistry {
    let mut r = McpRegistry::new();
    r.add_tool(Box::new(FnMcpTool::new(
        "add",
        "Adds two numbers",
        json!({ "type": "object", "properties": { "a": {"type":"number"}, "b": {"type":"number"} } }),
        |args| {
            let a = args.get("a").and_then(|v| v.as_i64()).unwrap_or(0);
            let b = args.get("b").and_then(|v| v.as_i64()).unwrap_or(0);
            Ok(json!({ "sum": a + b }))
        },
    )));
    r.add_tool(Box::new(FnMcpTool::new(
        "boom",
        "Always errors",
        json!({}),
        |_| Err(McpToolError("kaboom".to_string())),
    )));
    r.add_resource_provider(Box::new(
        InMemoryResourceProvider::new("vault://")
            .add("vault://note/1", "First note", "text/plain", "hello world"),
    ));
    r
}

fn rpc(method: &str, params: serde_json::Value, id: i64) -> serde_json::Value {
    json!({ "jsonrpc": "2.0", "id": id, "method": method, "params": params })
}

#[test]
fn initialize_returns_protocol_version() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r.dispatch(&rpc("initialize", json!({}), 1), &info).unwrap();
    assert_eq!(resp["result"]["protocolVersion"], "2024-11-05");
    assert_eq!(resp["result"]["serverInfo"]["name"], "circleai-mcp");
    // The id is stringified per the C# `id?.ToJsonString()`.
    assert_eq!(resp["id"], "1");
}

#[test]
fn tools_list_returns_registered_tools() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r.dispatch(&rpc("tools/list", json!({}), 2), &info).unwrap();
    let tools = resp["result"]["tools"].as_array().unwrap();
    assert_eq!(tools.len(), 2);
    assert_eq!(tools[0]["name"], "add");
}

#[test]
fn tools_call_success_wraps_content() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r
        .dispatch(&rpc("tools/call", json!({ "name": "add", "arguments": { "a": 2, "b": 3 } }), 3), &info)
        .unwrap();
    assert_eq!(resp["result"]["isError"], false);
    let text = resp["result"]["content"][0]["text"].as_str().unwrap();
    // The tool result is JSON-serialised into the text field.
    let inner: serde_json::Value = serde_json::from_str(text).unwrap();
    assert_eq!(inner["sum"], 5);
}

#[test]
fn tools_call_tool_error_sets_is_error() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r
        .dispatch(&rpc("tools/call", json!({ "name": "boom", "arguments": {} }), 4), &info)
        .unwrap();
    assert_eq!(resp["result"]["isError"], true);
    assert_eq!(resp["result"]["content"][0]["text"], "kaboom");
}

#[test]
fn tools_call_unknown_tool_errors() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r
        .dispatch(&rpc("tools/call", json!({ "name": "nope" }), 5), &info)
        .unwrap();
    assert_eq!(resp["error"]["code"], -32602);
}

#[test]
fn tools_call_missing_name_errors() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r
        .dispatch(&rpc("tools/call", json!({}), 6), &info)
        .unwrap();
    assert_eq!(resp["error"]["code"], -32602);
}

#[test]
fn resources_list_and_read() {
    let r = registry();
    let info = McpServerInfo::default();

    let list = r.dispatch(&rpc("resources/list", json!({}), 7), &info).unwrap();
    let resources = list["result"]["resources"].as_array().unwrap();
    assert_eq!(resources.len(), 1);
    assert_eq!(resources[0]["uri"], "vault://note/1");
    // Description defaults to the name when None.
    assert_eq!(resources[0]["description"], "First note");

    let read = r
        .dispatch(&rpc("resources/read", json!({ "uri": "vault://note/1" }), 8), &info)
        .unwrap();
    assert_eq!(read["result"]["contents"][0]["text"], "hello world");
}

#[test]
fn resources_read_no_provider_for_scheme() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r
        .dispatch(&rpc("resources/read", json!({ "uri": "models://x" }), 9), &info)
        .unwrap();
    assert_eq!(resp["error"]["code"], -32602);
    assert!(resp["error"]["message"].as_str().unwrap().contains("No provider"));
}

#[test]
fn resources_read_not_found() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r
        .dispatch(&rpc("resources/read", json!({ "uri": "vault://note/999" }), 10), &info)
        .unwrap();
    assert!(resp["error"]["message"].as_str().unwrap().contains("Resource not found"));
}

#[test]
fn unknown_method_returns_method_not_found() {
    let r = registry();
    let info = McpServerInfo::default();
    let resp = r.dispatch(&rpc("frobnicate", json!({}), 11), &info).unwrap();
    assert_eq!(resp["error"]["code"], -32601);
}

#[test]
fn notifications_initialized_returns_none() {
    let r = registry();
    let info = McpServerInfo::default();
    let req = json!({ "jsonrpc": "2.0", "method": "notifications/initialized" });
    assert!(r.dispatch(&req, &info).is_none());
}

#[test]
fn missing_jsonrpc_version_is_invalid_request() {
    let r = registry();
    let info = McpServerInfo::default();
    let req = json!({ "id": 1, "method": "tools/list" });
    let resp = r.dispatch(&req, &info).unwrap();
    assert_eq!(resp["error"]["code"], -32600);
}

#[test]
fn dispatch_body_handles_batch() {
    let r = registry();
    let info = McpServerInfo::default();
    let body = json!([
        rpc("tools/list", json!({}), 1),
        { "jsonrpc": "2.0", "method": "notifications/initialized" },
        rpc("initialize", json!({}), 2),
    ])
    .to_string();
    let out = r.dispatch_body(&body, &info);
    let arr = out.as_array().unwrap();
    // The notification produces no response → 2 entries.
    assert_eq!(arr.len(), 2);
}

#[test]
fn dispatch_body_parse_error() {
    let r = registry();
    let info = McpServerInfo::default();
    let out = r.dispatch_body("{not json", &info);
    assert_eq!(out["error"]["code"], -32700);
}
